### PR TITLE
Additions for the GEM SliceTest

### PR DESCRIPTION
**Pull request by calabria**
This PR introduces the needed changes to have a 2017 scenario with the GEM chambers foreseen for the next year slice test. We define a specific scenario, Extended2017Muon, where the simulation/reconstruction stops at local reco. We would like to have this in 900_pre2, because it is enough for the moment to be able to validate the new scenario with the relVal production foreseen for that pre-release, even though we don't enter the default 2017 scenario.

@jshlee @pietverwilligen 

schermata 2016-12-06 alle 18 55 42

### CODE REVIEW

File: Configuration/StandardSequences/python/Eras.py [hunk 1]
```diff
@@ -20,6 +20,7 @@ def __init__(self):
                  'Run2_2016_trackingLowPU',
                  'Run2_2016_pA',
                  'Run2_2017',
+                 'Run2_2017Muon',
```
Comment (makortel) [inline Configuration/StandardSequences/python/Eras.py]: I believe you should add `run2_GEMSliceTest` to `internalUseMods` below (but I can't place this comment there).

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by ianna** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by cvuosalo** [comment]
@calabria: I tried to run one of the new workflows, 11211.0_TenMuExtendedE_0_200+TenMuExtendedE_0_200_pythia8_2017MuonPU, but it failed in step2 with the error below. Did I do something wrong? Is there another new workflow that runs successfully?
```
----- Begin Fatal Exception 08-Dec-2016 22:17:28 CET-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=MixingModule label='mix'
Exception Message:
RootEmbeddedFileSequence no input files specified for secondary input source.
----- End Fatal Exception -------------------------------------------------
```

**Comment by makortel** [comment]
@cvuosalo "Known" "feature" of runTheMatrix when the `defaultDatasets` in `relval_steps.py` are not specified (as in this case). Better test with noPU workflows.

IMHO the machinery could be enhanced to give a meaningful error already before generating the configs (though I don't know how difficult this would be). OTOH sometimes it could be useful to generate the PU configs nevertheless and "do something" for the pileup input if one desires to run them.

**Comment by wddgit** [comment]
It could be totally unrelated to this case, but I have sometimes seen an error similar to this if I have not obtained my grid certificate ... I know nothing about this PR.
voms-proxy-init --rfc --voms cms

**Comment by slava77** [comment]
On 12/8/16 1:25 PM, Carl Vuosalo wrote:
> @calabria <https://github.com/calabria>: I tried to run one of the new
> workflows,
> 11211.0_TenMuExtendedE_0_200+TenMuExtendedE_0_200_pythia8_2017MuonPU,
> but it failed in step2 with the error below. Did I do something wrong?
> Is there another new workflow that runs successfully?


As already suggested, use a workflow without PU.

**Comment by cvuosalo** [comment]
+1

For #16900 5078abf086191ca640fbafbe62d547d627a7e6d9

Adding a new scenario for GEM slice testing in 2017. There should be no change in existing workflows.

The code changes are satisfactory, and Jenkins tests against baseline CMSSW_9_0_X_2016-12-07-1100 show no significant differences, as expected. A check of the expanded configs for workflows 10021.0_TenMuE_0_200+TenMuE_0_200_pythia8_2017 and 21211.0_TenMuExtendedE_0_200+TenMuExtendedE_0_200_pythia8_2023D4 against baseline CMSSW_9_0_X_2016-12-06-0800 shows no differences at all, so this PR is not changing existing upgrade workflows. A test of new workflow 11021.0_TenMuE_0_200+TenMuE_0_200_pythia8_2017MuonDesign ran successfully through all steps.

**Comment by Dr15Jones** [comment]
+1
Change to IOMC/RandomEngine is fine

**Comment by davidlange6** [comment]
Hi @calabria - please rebase this PR - it no longer merges.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by ianna** [comment]
+1

**Comment by cvuosalo** [comment]
+1

For #16900 88c76b1cf9c358f2d781f40ab6c254b7a2a03402

Second approval, after re-base. New Jenkins tests still show no problems.

**Comment by dmitrijus** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by davidlange6** [comment]
Hi @calabria - there is a merge conflict here.

Otherwise, what prevents this from being part of the mainstream 2017 geometry?

**Comment by calabria** [comment]
Hi @davidlange6 sorry I am not sure I uderstood what you mean: are you asking me to rebase? from our side there is no issue that prevents us from entering the mainstream geometry if we are allowed to do it. We simply did not want to interfere with the MC production. If now that 810 is out we are allowed to enter Extended2017 I will modify the PR.

**Comment by davidlange6** [comment]
> On Dec 21, 2016, at 7:07 AM, Cesare <notifications@github.com> wrote:
> 
> Hi @davidlange6 sorry I am not sure I uderstood what you mean: are you asking me to rebase? from our side there is no issue that prevents us from entering the mainstream geometry if we are allowed to do it. We simply did not want to interfere with the MC production. If now that 810 is out we are allowed to enter Extended2017 I will modify the PR.
> 

its 90x so it by definition can not interfere with 810…

but yes, this PR needs a rebase in addition.

**Comment by slava77** [comment]
On 12/21/16 4:12 AM, David Lange wrote:
> 
>> On Dec 21, 2016, at 7:07 AM, Cesare <notifications@github.com> wrote:
>> 
>> Hi @davidlange6 sorry I am not sure I uderstood what you mean: are you asking me to rebase? from our side there is no issue that prevents us from entering the mainstream geometry if we are allowed to do it. We simply did not want to interfere with the MC production. If now that 810 is out we are allowed to enter Extended2017 I will modify the PR.
>> 
> 
> its 90x so it by definition can not interfere with 810…
> 
> but yes, this PR needs a rebase in addition.


This is different from a rather firm message we got from you about a
month ago
that geometry is to be frozen in 810 for the purposes of 2017 MC.
Has this changed?

**Comment by davidlange6** [comment]
> On Dec 21, 2016, at 9:19 AM, Slava Krutelyov <notifications@github.com> wrote:
> 
> On 12/21/16 4:12 AM, David Lange wrote:
> > 
> >> On Dec 21, 2016, at 7:07 AM, Cesare <notifications@github.com> wrote:
> >> 
> >> Hi @davidlange6 sorry I am not sure I uderstood what you mean: are you asking me to rebase? from our side there is no issue that prevents us from entering the mainstream geometry if we are allowed to do it. We simply did not want to interfere with the MC production. If now that 810 is out we are allowed to enter Extended2017 I will modify the PR.
> >> 
> > 
> > its 90x so it by definition can not interfere with 810…
> > 
> > but yes, this PR needs a rebase in addition.
> 
> 
> This is different from a rather firm message we got from you about a
> month ago
> that geometry is to be frozen in 810 for the purposes of 2017 MC.
> Has this changed?
> 

for810 and thus SIM (in the current plan) The geometry has already evolved in 90x with the gem slice (adding new stuff so does not mean real inconsistencies) so I don’t think I’m saying anything new.

https://github.com/cms-sw/cmssw/pull/16601

**Comment by slava77** [comment]
On 12/21/16 6:23 AM, David Lange wrote:
> 
>> On Dec 21, 2016, at 9:19 AM, Slava Krutelyov <notifications@github.com> wrote:
>> 
>> On 12/21/16 4:12 AM, David Lange wrote:
>> > 
>> >> On Dec 21, 2016, at 7:07 AM, Cesare <notifications@github.com> wrote:
>> >> 
>> >> Hi @davidlange6 sorry I am not sure I uderstood what you mean: are you asking me to rebase? from our side there is no issue that prevents us from entering the mainstream geometry if we are allowed to do it. We simply did not want to interfere with the MC production. If now that 810 is out we are allowed to enter Extended2017 I will modify the PR.
>> >> 
>> > 
>> > its 90x so it by definition can not interfere with 810…
>> > 
>> > but yes, this PR needs a rebase in addition.
>> 
>> 
>> This is different from a rather firm message we got from you about a
>> month ago
>> that geometry is to be frozen in 810 for the purposes of 2017 MC.
>> Has this changed?
>> 
> 
> for810 and thus SIM (in the current plan) The geometry has already
> evolved in 90x with the gem slice (adding new stuff so does not mean
> real inconsistencies) so I don’t think I’m saying anything new.
> 
> https://github.com/cms-sw/cmssw/pull/16601


Sorry, I couldn't parse your message:
you mention #16601
which is "new scenario for 2017 (not disturbing the present default)"

But the Q&A earlier with Cesare imply that you suggest to integrate this
geometry into the default 2017 geometry.

**Comment by davidlange6** [comment]
> On Dec 21, 2016, at 9:35 AM, Slava Krutelyov <notifications@github.com> wrote:
> 
> On 12/21/16 6:23 AM, David Lange wrote:
> > 
> >> On Dec 21, 2016, at 9:19 AM, Slava Krutelyov <notifications@github.com> wrote:
> >> 
> >> On 12/21/16 4:12 AM, David Lange wrote:
> >> > 
> >> >> On Dec 21, 2016, at 7:07 AM, Cesare <notifications@github.com> wrote:
> >> >> 
> >> >> Hi @davidlange6 sorry I am not sure I uderstood what you mean: are you asking me to rebase? from our side there is no issue that prevents us from entering the mainstream geometry if we are allowed to do it. We simply did not want to interfere with the MC production. If now that 810 is out we are allowed to enter Extended2017 I will modify the PR.
> >> >> 
> >> > 
> >> > its 90x so it by definition can not interfere with 810…
> >> > 
> >> > but yes, this PR needs a rebase in addition.
> >> 
> >> 
> >> This is different from a rather firm message we got from you about a
> >> month ago
> >> that geometry is to be frozen in 810 for the purposes of 2017 MC.
> >> Has this changed?
> >> 
> > 
> > for810 and thus SIM (in the current plan) The geometry has already
> > evolved in 90x with the gem slice (adding new stuff so does not mean
> > real inconsistencies) so I don’t think I’m saying anything new.
> > 
> > https://github.com/cms-sw/cmssw/pull/16601
> 
> 
> Sorry, I couldn't parse your message:
> you mention #16601
> which is "new scenario for 2017 (not disturbing the present default)"
> 
> But the Q&A earlier with Cesare imply that you suggest to integrate this
> geometry into the default 2017 geometry.

yes, you are right about the PR - I think I missed that at the time. Anyway, whats the downside to changing the geometry? You’ll presumably need this in the reco in any case.. Likely easier to use the same reco geometry everywhere (data/mc) if we can..

**Comment by slava77** [comment]
On 12/21/16 6:39 AM, David Lange wrote:
>> > for810 and thus SIM (in the current plan) The geometry has already
>> > evolved in 90x with the gem slice (adding new stuff so does not mean
>> > real inconsistencies) so I don’t think I’m saying anything new.
>> > 
>> > https://github.com/cms-sw/cmssw/pull/16601
>> 
>> 
>> Sorry, I couldn't parse your message:
>> you mention #16601
>> which is "new scenario for 2017 (not disturbing the present default)"
>> 
>> But the Q&A earlier with Cesare imply that you suggest to integrate this
>> geometry into the default 2017 geometry.
> 
> yes, you are right about the PR - I think I missed that at the time.
> Anyway, whats the downside to changing the geometry? You’ll presumably
> need this in the reco in any case.. Likely easier to use the same reco
> geometry everywhere (data/mc) if we can..



Is SIM for 2017 going to be done with 810 or with 900?


For the reco, minimally, I expect that this slice test data enters the
standard workflows as the pixel pilot blade data did (or rather how it
did not).

**Comment by davidlange6** [comment]
> On Dec 21, 2016, at 9:43 AM, Slava Krutelyov <notifications@github.com> wrote:
> 
> On 12/21/16 6:39 AM, David Lange wrote:
> >> > for810 and thus SIM (in the current plan) The geometry has already
> >> > evolved in 90x with the gem slice (adding new stuff so does not mean
> >> > real inconsistencies) so I don’t think I’m saying anything new.
> >> > 
> >> > https://github.com/cms-sw/cmssw/pull/16601
> >> 
> >> 
> >> Sorry, I couldn't parse your message:
> >> you mention #16601
> >> which is "new scenario for 2017 (not disturbing the present default)"
> >> 
> >> But the Q&A earlier with Cesare imply that you suggest to integrate this
> >> geometry into the default 2017 geometry.
> > 
> > yes, you are right about the PR - I think I missed that at the time.
> > Anyway, whats the downside to changing the geometry? You’ll presumably
> > need this in the reco in any case.. Likely easier to use the same reco
> > geometry everywhere (data/mc) if we can..
> 
> 
> 
> Is SIM for 2017 going to be done with 810 or with 900?

afaik, 810 is the current plan. But we can always be ready in case 900 is desired for other reasons (such as a further pixel material budget change or a descope of part of the hcal upgrade, etc etc)

File: L1Trigger/CSCTriggerPrimitives/python/cscTriggerPrimitiveDigis_cfi.py [hunk 2]
```diff
@@ -510,6 +510,17 @@
...
 ## GEM-CSC ILT in ME1/1
+from Configuration.Eras.Modifier_run2_GEMSliceTest_cff import run2_GEMSliceTest
+run2_GEMSliceTest.toModify( cscTriggerPrimitiveDigis,
+                        GEMPadDigiProducer = cms.InputTag("simMuonGEMPadDigis"),
+                        commonParam = dict(
+                            isSLHC = cms.bool(True),
+                            smartME1aME1b = cms.bool(True),
+                            runME11ILT = cms.bool(True)),
+                        clctSLHC = dict(clctNplanesHitPattern = 3),
+                        me11tmbSLHCGEM = me11tmbSLHCGEM
```
Comment (slava77) [inline L1Trigger/CSCTriggerPrimitives/python/cscTriggerPrimitiveDigis_cfi.py]: is this what we will actually have in 2017 TMB firmware or is this just a wishful thinking simulation?
@dildick

**Comment by slava77** [comment]
hold

On 12/21/16 6:48 AM, David Lange wrote:
> 
>> On Dec 21, 2016, at 9:43 AM, Slava Krutelyov <notifications@github.com> wrote:
>> 
>> On 12/21/16 6:39 AM, David Lange wrote:
>> >> > for810 and thus SIM (in the current plan) The geometry has already
>> >> > evolved in 90x with the gem slice (adding new stuff so does not mean
>> >> > real inconsistencies) so I don’t think I’m saying anything new.
>> >> > 
>> >> > https://github.com/cms-sw/cmssw/pull/16601
>> >> 
>> >> 
>> >> Sorry, I couldn't parse your message:
>> >> you mention #16601
>> >> which is "new scenario for 2017 (not disturbing the present default)"
>> >> 
>> >> But the Q&A earlier with Cesare imply that you suggest to integrate this
>> >> geometry into the default 2017 geometry.
>> > 
>> > yes, you are right about the PR - I think I missed that at the time.
>> > Anyway, whats the downside to changing the geometry? You’ll presumably
>> > need this in the reco in any case.. Likely easier to use the same reco
>> > geometry everywhere (data/mc) if we can..
>> 
>> 
>> 
>> Is SIM for 2017 going to be done with 810 or with 900?
> 
> afaik, 810 is the current plan. But we can always be ready in case 900
> is desired for other reasons (such as a further pixel material budget
> change or a descope of part of the hcal upgrade, etc etc)
> 
> 

OK.
So, it sounds like the action items are
- integrate GEM slice test in the standard (DB) geometry for 2017
[separate PR and a GT, apparently]
- update this PR to use run2_gemSliceTest in Run2_2017 era instead of
the specialized one.

should we take out reco "+1"?

**Comment by davidlange6** [comment]
even if the reco objects are intended to be separated out of the main workflow - if we can have one geometry, life will be easier (but maybe its impossible)

**Comment by calabria** [comment]
@slava77 am I supposed to take care about point 1? If so, can you please give some hints on how to proceed? Concerning point 2, should I drop the specialized era and scenario?

**Comment by slava77** [comment]
@calabria 
you or, most likely, @bsunanda or @ianna will need to update the geometry setup for 2017 default to include the GEM slice test.

Once the update shows up in the GT, you can activate relevant parts of DIGI and RECO from this PR.

File: L1Trigger/CSCTriggerPrimitives/python/cscTriggerPrimitiveDigis_cfi.py (hunk 2 above)
Comment (dildick) [inline L1Trigger/CSCTriggerPrimitives/python/cscTriggerPrimitiveDigis_cfi.py]: I doubt that a fully working integrated trigger will be implemented and tested by the slice test. I will check with the UCLA colleagues.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by ianna** [comment]
@slava77 - just to confirm: 
- the default 2017 scenario to be updated to include the GEM slice
- 2017Muon scenario to be deleted

I'll submit a separate PR for that.

**Comment by ianna** [comment]
@calabria and @slava77 - FYI, https://github.com/cms-sw/cmssw/pull/17115

The DB tag and GT request will follow shortly (after the tests are completed).

**Comment by calabria** [comment]
thanks @ianna !

**Comment by ianna** [comment]
@calabria - here is what I see as a GEM reco geometry read from DB:
screen shot 2017-01-06 at 15 07 51
To read it directly from a GEMRecoGeometryRcd DB tag (e.g GT when it's ready) the following configuration should be used:
```
from Geometry.GEMGeometryBuilder.gemGeometryDB_cfi import *
```
I will update 2017 RecoDB configuration fragment in a separate PR.

**Comment by ianna** [comment]
@calabria - now you should be able to use Run2_2017 to pick up GEM reco geometry from DB:
```
process = cms.Process('YOUR_PROCESS_NAME',eras.Run2_2017)
```

**Comment by calabria** [comment]
@ianna yes thanks, I will update this PR asap

**Comment by slava77** [comment]
@calabria 
please clarify on the status of updates for this PR (or related to it).
Thank you.

**Comment by calabria** [comment]
@slava77 I am working right now on it. I am not done yet

**Comment by kpedro88** [comment]
@calabria there still appear to be merge conflicts

**Comment by kpedro88** [comment]
okay

**Comment by slava77** [comment]
@calabria 
there is a merge conflict in  Configuration/StandardSequences/python/Eras.py 
please resolve

**Comment by calabria** [comment]
I am stuck with git, I don't know how to proceed. Maybe it is better to close this PR

**Comment by slava77** [comment]
https://twiki.cern.ch/twiki/bin/viewauth/CMS/SWGuideHATS2016Git#CMSSW_rebasing_recipe
in case you haven't tried it already

**Comment by calabria** [comment]
yes I know about that twiki but I don't remember what was the initial IB... so I guess I have to use git rebase --onto NEW_IB start end
but what is end in this case?

**Comment by slava77** [comment]
On 2/8/17 7:04 AM, Cesare wrote:
> yes I know about that twiki but I don't remember what was the initial
> IB... so I guess I have to use git rebase --onto NEW_IB start end
> but what is end in this case?

you can use the commit id, not the tag
git merge-base CMSSW_9_0_X

**Comment by ianna** [comment]
please test

**Comment by calabria** [comment]
no please wait, I have not been able to solve the problems with git

**Comment by slava77** [comment]
@calabria 
maybe if you are stuck with git it's easier just to start with a new branch and commits 
(and then either open a new PR or push --force into this PR topic branch to overwrite)

**Comment by slava77** [comment]
the error in the last build was 
```
  File "src/SimMuon/Configuration/python/SimMuon_cff.py", line 45
    from from Configuration.Eras.Modifier_run2_GEM_2017_cff import run2_GEM_2017
            ^
SyntaxError: invalid syntax
```
seems like something easy to fix

**Comment by calabria** [comment]
@slava77 thanks. yes I noticed it and fixed it... But I am preparing a new branch, I don't think I can continue with this one... Unfortunately the validation packages are now crashing for some reason and I am trying to understand why...

**Comment by franzoni** [comment]
hello @calabria 
I gather this instance of this PR is no longer relevant
might you want to close ?

**Comment by slava77** [comment]
@calabria 
Should we still expect a PR with GEM slice test?

If I understand correctly, current default 2017 SIM has the slice test chambers
@ianna please confirm

If that's the case, addition of GEM local reco should be possible in 91X.

**Comment by calabria** [comment]
Hi @slava77  yes sure, but unfortunately we have another issue to solve: we are having problems with the GEMGeometry taken from the Database, when asking the regions of the Geometry we get a crash. Maybe the geometry constructed from the database is not as complete as the geometry we can take from DDD. We realized the part where chambers, rings, stations and regions are created is missing in the geometryFromCondDB. We have to work on an increased description of the CondDB geometry first. We already contacted @ianna to get some help on how to proceed because it is not perfectly clear: we need to know how we can test, if we need to update the payload...

**Comment by ianna** [comment]
@slava77 - yes, GEM slice test chambers are in 2017 SIM

@calabria - how can I reproduce the crash? Thanks

**Comment by calabria** [comment]
@ianna the problem shows up in the validation, if I run the simulation up to the reconstruction stepeverything "seems" to work fine. So in order to reproduce the crash it is enough if you run the reco step + validation, so if you run 10411 you should see the problem.

**Comment by ianna** [comment]
@calabria - thanks, I can reproduce the crash:
```
#5  0x00007f3051ab317a in GEMBaseValidation::initGeometry(edm::EventSetup const&) () from /build/yana/GEM/CMSSW_9_0_X_2017-02-22-2300/lib/slc6_amd64_gcc530/libValidationMuonGEMHits.so
#6  0x00007f3051aac43a in GEMHitsValidation::bookHistograms(DQMStore::IBooker&, edm::Run const&, edm::EventSetup const&) () from /build/yana/GEM/CMSSW_9_0_X_2017-02-22-2300/lib/slc6_amd64_gcc530/libValidationMuonGEMHits.so
#7  0x00007f3095c0281a in DQMEDAnalyzer::beginRun(edm::Run const&, edm::EventSetup const&) () from /cvmfs/cms-ib.cern.ch/nweek-02460/slc6_amd64_gcc530/cms/cmssw/CMSSW_9_0_X_2017-02-22-1100/lib/slc6_amd64_gcc530/libDQMServicesCore.so
```
The vector of GEMRegions is empty, so the code accessing its constituents is not safe.

GEM geometry builder from DB constructs only Eta partitions (80 of them in this particular case): http://cmslxr.fnal.gov/source/Geometry/GEMGeometryBuilder/src/GEMGeometryBuilderFromCondDB.cc#0033

Usually, RECO geometry contains only sensitive volumes, however, if GEM reco geometry needs to provide more (regions, superChambers and chambers) GEMGeometryBuilderFromCondDB is the place to add these.

Please, let me know if you need my help in implementing the code.

**Pull request closed without merging**